rewrite-core/src/main/java/org/openrewrite/scm/AzureDevOpsScm.java

```diff
@@ -15,7 +15,16 @@
  */
 package org.openrewrite.scm;
 
+import org.openrewrite.internal.lang.Nullable;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
 public class AzureDevOpsScm implements Scm {
+    private static final Pattern AZURE_DEVOPS_HTTP_REMOTE =
```

Comment:
I dont think these patterns cover every possible url.
for example, this url is technically also valid: 
`ssh://ssh.dev.azure.com:22/v3/org/project/repo`
 (and such test cases are there for `UrlComponents`)
 
 I like the other changed, but using matchers here, and not for other SCM's means we have 2 ways of doing things and we should avoid that. Since we cannot use matchers for all Scm's I think we should stick with parsing.